src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs

```diff
@@ -0,0 +1,245 @@
+        }
+
+        private static string GetResponseHeaderStringInfo(SafeWinHttpHandle requestHandle, uint infoLevel)
+        {
+            uint bytesNeeded = 0;
+            bool results = false;
+
+            // Call WinHttpQueryHeaders once to obtain the size of the buffer needed.  The size is returned in
+            // bytes but the API actually returns Unicode characters.
+            if (!Interop.WinHttp.WinHttpQueryHeaders(
+                requestHandle,
+                infoLevel,
+                Interop.WinHttp.WINHTTP_HEADER_NAME_BY_INDEX,
+                null,
+                ref bytesNeeded,
+                IntPtr.Zero))
+            {
+                int lastError = Marshal.GetLastWin32Error();
+                if (lastError == Interop.WinHttp.ERROR_WINHTTP_HEADER_NOT_FOUND)
+                {
+                    return null;
+                }
+
+                if (lastError != Interop.WinHttp.ERROR_INSUFFICIENT_BUFFER)
+                {
+                    throw WinHttpException.CreateExceptionUsingError(lastError);
+                }
+            }
+
+            // Allocate space for the buffer.
+            int charsNeeded = (int)bytesNeeded / 2;
+            var buffer = new StringBuilder(charsNeeded, charsNeeded);
+
+            results = Interop.WinHttp.WinHttpQueryHeaders(
+                requestHandle,
+                infoLevel,
+                Interop.WinHttp.WINHTTP_HEADER_NAME_BY_INDEX,
+                buffer,
+                ref bytesNeeded,
+                IntPtr.Zero);
+            if (!results)
+            {
+                WinHttpException.ThrowExceptionUsingLastError();
+            }
+
+            return buffer.ToString();
+        }
+
+        private static void ParseResponseHeaders(
+            SafeWinHttpHandle requestHandle,
+            HttpResponseMessage response,
+            bool stripEncodingHeaders)
+        {
+            string rawResponseHeaders = GetResponseHeaderStringInfo(
+                requestHandle,
+                Interop.WinHttp.WINHTTP_QUERY_RAW_HEADERS_CRLF);
+            string[] responseHeaderArray = rawResponseHeaders.Split(
+                s_HttpHeadersSeparator,
+                StringSplitOptions.RemoveEmptyEntries);
+
+            // Parse the array of headers and split them between Content headers and Response headers.
+            // Skip the first line which contains status code, etc. information that we already parsed.
+            for (int i = 1; i < responseHeaderArray.Length; i++)
+            {
+                int colonIndex = responseHeaderArray[i].IndexOf(':');
+
+                // Skip malformed header lines that are missing the colon character.
+                if (colonIndex > 0)
+                {
+                    string headerName = responseHeaderArray[i].Substring(0, colonIndex);
+                    string headerValue = responseHeaderArray[i].Substring(colonIndex + 1).Trim(); // Normalize header value by trimming white space.
```

Comment:
You might consider doing something like the SubstringTrim here (https://github.com/dotnet/corefx/blob/c02d33b18398199f6acc17d375dab154e9a1df66/src/System.Net.WebHeaderCollection/src/System/Net/HeaderInfoTable.cs#L108) rather than doing Substring and then Trim.